### DISCUSSION TITLE
More optimal SPI communication, for Wiznet W5500 and possibly others.

### DISCUSSION THREAD

**Discussion by Pe3ucTop** [cesanta/mongoose#2925] [Feature requests]
Hello everyone, 

I start using library and like it, great job.
But, I noticed in case of using SPI-Ethernet adapter (Wiznet W5500 in my case) there is ineffective SPI communication and I'm ready to prepare pull request with improvements.
Description of ineffective communication and limitations of library used function  `uint8_t (*txn)(void *, uint8_t);  // SPI transaction: write 1 byte, read reply` :
- it require system support bidirectional SPI transactions, in many cases not needed extra data operations, because we either transmit packet or receive.
- it operate with single byte, in loop by CPU
-- which limit use of existing resources like DMA or FIFO buffers
-- which force to start/stop SPI peripheral for each byte, a bit insane
- it limit of use extra SPI features like unidirectional dual/quad line transfers (not needed for W5500 , but could be used for SPI Flash / SRAM / etc)

I modify my copy to work with unidirectional functions
`void (*tx)(void *, uint8_t *buf, uint32_t len);  // SPI transaction: write N bytes from provided buffer`
`void (*rx)(void *, uint8_t *buf, uint32_t len);  // SPI transaction: read N bytes to provided buffer`
and it work perfect.
Or I can prepare single universal function usage, which could be any directional (uni/bi)
`void (*transfer)(void *, uint8_t *buf, uint32_t len, uint8_t direction);  // SPI transaction: write/read/rw N bytes from/to/replace buffer`
extra multi-line unidirectional transfers could be included here with `direction`  value.

Would like to see your thoughts and comments.

**Comment by scaprile** [comment]
Thanks, we are not interested

**Reply by Maksons** [reply]
Thanks for response and it's sad to hear that optimizations are not interested.
Is there any particular reason why this improvement got rejected?

**Reply by scaprile** [reply]
No particular reasons, just differences in criteria. We do optimize where we need to.
